DisguiseUnityRenderStream/Runtime/Texture2DPool.cs

```diff
@@ -0,0 +1,210 @@
+using System;
+using System.Collections.Generic;
+using Disguise.RenderStream.Utils;
+using UnityEngine;
+using UnityEngine.PlayerLoop;
+using UnityEngine.Pool;
+
+namespace Disguise.RenderStream
+{
+    public struct Texture2DDescriptor : IEquatable<Texture2DDescriptor>
+    {
+        public int Width;
+        public int Height;
+        public TextureFormat Format;
+        public bool Linear;
+
+        /// <summary>
+        /// Disguise uses a black 1x1 placeholder texture.
+        /// It's bound initially and during input parameter swapping.
+        /// We treat it as a persistent texture.
+        /// </summary>
+        public bool IsPlaceholderTexture => Width == 1 && Height == 1;
+        
+        public override bool Equals(object obj)
+        {
+            return obj is Texture2DDescriptor other && Equals(other);
+        }
+        
+        public override int GetHashCode()
+        {
+            return HashCode.Combine(
+                Width,
+                Height,
+                (int)Format,
+                Linear
+            );
+        }
+
+        public bool Equals(Texture2DDescriptor other)
+        {
+            return
+                Width == other.Width &&
+                Height == other.Height &&
+                Format == other.Format &&
+                Linear == other.Linear;
+        }
+            
+        public static bool operator ==(Texture2DDescriptor lhs, Texture2DDescriptor rhs) => lhs.Equals(rhs);
+
+        public static bool operator !=(Texture2DDescriptor lhs, Texture2DDescriptor rhs) => !(lhs == rhs);
+
+        public override string ToString()
+        {
+            return $"{Width}x{Height} Format {Format} {(Linear ? "Linear" : "SRGB")}";
+        }
+    }
+
+    /// <summary>
+    /// Provides texture re-use across a frame and texture lifetime management.
+    /// Similar to <see cref="RenderTexture.GetTemporary(RenderTextureDescriptor)"/>.
+    ///
+    /// <remarks>Lifetime doesn't grow during frames where no textures from the pool were used.</remarks>
+    /// </summary>
+    public class Texture2DPool
```

Comment:
I'm really confused about what `Texture2DPool` is supposed to do, because it doesn't behave like typical pool.

* I call `Get()`. Assuming this is a fresh pool, a new texture is created.
* Let's say I call `Get()` again. If the descriptor is the same, I get the same texture I received from the first `Get()`. Huh?
* The textures get automatically destroyed after some number of frames. There's no way for the user to manually release textures back into the pool. Each call to `Get()` with a certain descriptor resets the countdown. After the the texture is automatically "released" it's destroyed for good. If I try to `Get()` again with that descriptor, I have to pay the cost of creating a new texture.

So why does it behave like this? Why can't I `Get` more than one texture of the same type? Why can't I return anything to the pool?

Reply:
True this pooling logic wouldn't work for output textures. It works for input textures where we do the following for every Disguise input image:
1. Prepare a "scratch" Texture2D that matches the Disguise image parameters
2. Read the Disguise image into the scratch texture
3. Blit (while flipping Y) the scratch texture into the target RenderTexture asset

In step 1 we can re-use a previously created scratch texture when the Disguise image parameters (descriptors) are identical. This is the `Get` logic.

We go through all these steps for every input texture each time there's a `HasNewFrameData`. So if a certain texture's descriptor hasn't been requested in a while then we know it's not being used by any Disguise input images anymore. This is the release logic.

I could rename it to `DisguiseInputTexture2DPool` to signal that it's for a specific use case.

Reply:
OK, that makes sense. Maybe call it something like `TemporaryTexture2DManager` or something like that 🙂